service/controller/v26/resource/cleanupendpointips/create.go

```diff
@@ -0,0 +1,182 @@
+			Logger:     r.logger,
+			RestConfig: restConfig,
+		}
+		k8sClients, err := k8sclient.NewClients(clientsConfig)
+		if tenant.IsAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "tenant cluster is not available")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		k8sClient = k8sClients.K8sClient()
+		r.logger.LogCtx(ctx, "level", "debug", "message", "created Kubernetes client for tenant cluster")
+	}
+
+	// We need to fetch the nodes being registered within the tenant cluster's
+	// Kubernetes API.
+	var nodes []corev1.Node
+	{
+		list, err := k8sClient.CoreV1().Nodes().List(metav1.ListOptions{})
+		if tenant.IsAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "tenant cluster is not available")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+		nodes = list.Items
+	}
+
+	// Fetch the list of pods running on the control plane. These pods serve VMs
+	// which in turn run the tenant cluster nodes.
+	var pods []corev1.Pod
+	{
+		n := key.ClusterID(customObject)
+		list, err := r.k8sClient.CoreV1().Pods(n).List(metav1.ListOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		pods = list.Items
+	}
+	// Check if all k8s-kvm pods in CP are registered as nodes in the TC.
+	if podsEqualNodes(pods, nodes) {
+		n := key.ClusterID(customObject)
+
+		{
+			masterEndpoint, err := r.k8sClient.CoreV1().Endpoints(n).Get(key.MasterID, metav1.GetOptions{})
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			epRemoved, masterEndpoint := removeDeadIPFromEndpoints(masterEndpoint, nodes)
+			if epRemoved > 0 {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("removing %d dead ips from the master endpoints", epRemoved))
+
+				_, err = r.k8sClient.CoreV1().Endpoints(n).Update(masterEndpoint)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+		}
+
+		{
+			workerEndpoint, err := r.k8sClient.CoreV1().Endpoints(n).Get(key.WorkerID, metav1.GetOptions{})
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			epRemoved, workerEndpoint := removeDeadIPFromEndpoints(workerEndpoint, nodes)
+			if epRemoved > 0 {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("removing %d dead ips from the worker endpoints", epRemoved))
+
+				_, err = r.k8sClient.CoreV1().Endpoints(n).Update(workerEndpoint)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// podsEqualNodes check if all k8s-kvm pods in CP are registered as nodes in the TC.
+func podsEqualNodes(pods []corev1.Pod, nodes []corev1.Node) bool {
+	if len(pods) != len(nodes) {
+		return false
+	}
+	for i := 0; i < len(pods); i++ {
+		// k8s api return result alphabetically sorted by name so we can simply compare indexes
```

Comment:
We should definitely not rely on this assumption of external dependencies. Tomorrow something breaks there or they change it and our systems start to act up. Imagine you have to debug this and figure out the reason. We can prevent this from happening right now.